### PR TITLE
Feature/voting

### PR DESCRIPTION
**Pull request by tnewell5**
Hey Travis!

This PR covers two following voting feature stories: #232 and #235
Let Claire or I know if you have any questions.

Thanks!
Tanya

### CODE REVIEW

File: test/components/idea_controls_test.js [hunk 1]
```diff
@@ -172,4 +183,63 @@ describe("<IdeaControls />", () => {
       })
     })
   })
+
+  describe("the vote button", () => {
+    context("when the stage is not idea-generation", () => {
```
Comment (vanderhoop) [inline test/components/idea_controls_test.js]: without looking at the implementation, i'd think that the only time we'd render the vote button is during the voting stage, right? does the category of the idea need to come into play, given that there won't be any action items under _after_ the voting stage is over?

File: test/components/vote_counter_test.js [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    category: "sad",
+    body: "redundant tests",
+    user_id: 1,
+    user: {
+      given_name: "Phil",
+    },
+    vote_count: 0,
+  }
+
+  it("renders a button that says Vote", () => {
```
Comment (vanderhoop) [inline test/components/vote_counter_test.js]: we can delete this test, as the "Vote" text is static

File: test/components/vote_counter_test.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    const voteCounter = shallow(
+      <VoteCounter
+        retroChannel={{}}
+        idea={idea}
+      />
+    )
+    const label = voteCounter.find("a")
+
+    expect(label).to.have.length(1)
+    expect(label.hasClass("pointing label")).to.be.true
```
Comment (vanderhoop) [inline test/components/vote_counter_test.js]: let's just have one expectation here, that the test of the label matches the idea's current vote count. and no need to specify that it's a pointing label

File: web/static/js/components/idea_controls.jsx [hunk 4]
```diff
@@ -21,6 +22,11 @@ const IdeaControls = props => {
   const highlightTitle = isHighlighted ? "De-Highlight Idea for Participants" : "Announce Idea to Channel"
 
   function renderIcons() {
+    if (stage !== "idea-generation" && category !== "action-item") {
```
Comment (vanderhoop) [inline web/static/js/components/idea_controls.jsx]: per the comment above, the condition here should just be whether it's the voting stage

File: web/static/js/components/idea_generation_lower_third.jsx [hunk 5]
```diff
@@ -12,10 +12,12 @@ const IdeaGenerationLowerThird = props => {
 
   const isFacilitator = currentUser.is_facilitator
   const stageConfig = stageConfigs[stage]
-  const showActionItem = stage !== "idea-generation"
+  const showActionItem = !["idea-generation", "voting"].includes(stage)
 
-  function wereActionItemsSubmitted() {
-    return showActionItem && !ideas.some(idea => idea.category === "action-item")
+  function progressionDisabled() {
```
Comment (vanderhoop) [inline web/static/js/components/idea_generation_lower_third.jsx]: why was this change necessary? it doesn't seem related to voting, but I'm wrong a lot?

File: web/static/js/services/retro_channel.js [hunk 6]
```diff
@@ -47,6 +47,10 @@ const applyListenerCallbacks = (retroChannel, store, actions) => {
     actions.deleteIdea(deletedIdea.id)
   })
 
+  retroChannel.on("vote_submitted", votedOnIdea => {
```
Comment (vanderhoop) [inline web/static/js/services/retro_channel.js]: let's add a test for this!

File: web/channels/retro_channel.ex [hunk 7]
```diff
@@ -76,6 +76,18 @@ defmodule RemoteRetro.RetroChannel do
     {:noreply, socket}
   end
 
+  def handle_in("submit_vote", %{"id" => id}, socket) do
+    idea = Repo.get(Idea, id)
```
Comment (vanderhoop) [inline web/channels/retro_channel.ex]: based on the handler i see in the retro, where we're only relying on `id` and `vote_count` attributes, there shouldn't be any need to preload the user here, which will save us a database hit. 

The select then update queries can be consolidated into an update query using the following syntax:

```elixir
    query = from i in Idea, where: i.id == ^id
    {_row_count, [updated_idea]} = Repo.update_all(query, [inc: [vote_count: 1]], returning: true)
```

Not the prettiest thing in the world, but gets the job done in one go.

File: web/static/js/components/idea_generation_lower_third.jsx (hunk 5 above)
Comment (vanderhoop) [inline web/static/js/components/idea_generation_lower_third.jsx]: let's hold off on this for this story, so that adding tests for itdoesn't hold up this PR

File: web/static/js/components/idea_board.jsx [hunk 8]
```diff
@@ -8,7 +8,7 @@ import styles from "./css_modules/idea_board.css"
 const IdeaBoard = props => {
   const { stage } = props
   const categories = ["happy", "sad", "confused"]
-  const showActionItem = stage !== "idea-generation"
+  const showActionItem = !["idea-generation", "voting"].includes(stage)
```
Comment (vanderhoop) [inline web/static/js/components/idea_board.jsx]: to avoid the `!` here, we could just say that we only want to show the action items during particular stages, rather than list the stages that we don't want, i.e. `["action-items", "closed"].includes(stage)`.

In the long term we want to extract these conditionals and have composed elements like `IdeaGenerationIdeaboard` and `ActionItemIdeaBoard`, as these in-component conditionals are only going to get more complex as we add stages.

Comment (vanderhoop) [inline web/static/js/components/idea_board.jsx]: also, once we've done away with the bang, let's update the tests for this condition (it only renders when the stage is either "action-items" or "action-item-distribution"

File: test/channels/retro_channel_test.exs [hunk 9]
```diff
@@ -172,4 +172,29 @@ defmodule RemoteRetro.RetroChannelTest do
...
+    end
+
+
+    @tag user: @mock_user
+    @tag idea: %Idea{category: "sad", body: "doggone keeper"}
+    test "results in the idea being updated in the database", %{socket: socket, idea: idea} do
+      idea_id = idea.id
+      push(socket, "submit_vote", %{id: idea_id})
+
+      :timer.sleep(50)
```
Comment (vanderhoop) [inline test/channels/retro_channel_test.exs]: not ideal, but i recall having to do something similar, due to the db insertion being fired off in another process by the channel

File: test/components/vote_counter_test.js (hunk 2 above)
Comment (tnewell5) [inline test/components/vote_counter_test.js]: done

File: test/components/vote_counter_test.js (hunk 3 above)
Comment (tnewell5) [inline test/components/vote_counter_test.js]: done

File: web/static/js/components/idea_generation_lower_third.jsx (hunk 5 above)
Comment (tnewell5) [inline web/static/js/components/idea_generation_lower_third.jsx]: This was a bug fix implemented per Sean's suggestion. He was under the impression that the progression button should be disabled until an idea has been submitted. Should we create a bug fix PR for this?

File: web/static/js/services/retro_channel.js (hunk 6 above)
Comment (tnewell5) [inline web/static/js/services/retro_channel.js]: will do

File: web/channels/retro_channel.ex (hunk 7 above)
Comment (tnewell5) [inline web/channels/retro_channel.ex]: will look into this

File: web/static/js/components/idea_board.jsx (hunk 8 above)
Comment (tnewell5) [inline web/static/js/components/idea_board.jsx]: okay

File: test/channels/retro_channel_test.exs (hunk 9 above)
Comment (tnewell5) [inline test/channels/retro_channel_test.exs]: noted

File: web/static/js/components/idea_generation_lower_third.jsx (hunk 5 above)
Comment (vanderhoop) [inline web/static/js/components/idea_generation_lower_third.jsx]: yes please!

File: test/components/idea_controls_test.js (hunk 1 above)
Comment (tnewell5) [inline test/components/idea_controls_test.js]: agreed to leave alone for now
